### ISSUE TITLE
SelectNodes not matching xpath where attribute name starts-with

### ISSUE THREAD

**Issue by Kenshinofkin** [zzzprojects/html-agility-pack#518]
Here is what to include in your request to make sure we implement a solution as quickly as possible. 

## 1. Description
SelectNodes is returning null when trying to get nodes where attribute name starts-with function is used. I tested the xpath and HTML on a couple of different sites and was able to get nodes back.

XPATH: `//*[@*[starts-with(name(), 'on')]]`
HTML: `<div onclick="alert('test');"></div>`
Expected Result: div should be returned from SelectNodes.

## 2. Exception
No Exception

```
Exception message:
Stack trace:
```

## 3. Fiddle or Project
[Fiddle](https://dotnetfiddle.net/autg8k)

## 4. Any further technical details

- HAP version: 1.11.53
- NET version: .net 7

Updated: Include correct xpath used and example fiddle.

**Comment by elgonzo** [comment]
While i am not one of the authors/maintainers of HtmlAgilityPack (i am just a user), i can tell that there are two issues with you writing an incorrect XPath expression.

**But there is also a bug in HtmlAgilityPack.** Before i address the incorrectness of your XPath expression, i want to address the HAP bug here first, as this is the only thing of interest for the HtmlAgilityPack maintainers. (_@Kenshinofkin  please keep the issue report therefore open, even though i suggest a workaround below._) 

For attribute nodes, HAP forgets to set the `HtmlNodeNavigator.Name` and only sets the `HtmlNodeNavigator.LocalName` property. This in turn makes the the XPath function `name()` misbehaving for attribute nodes. When HtmlNodeNavigator navigates to an attribute node, `HtmlNodeNavigator.Name` will not yield the attribute name but instead return the name of the last navigated element node, which is incorrect.

----------------------

With the HAP bug explained, lets now address your invalid XPath expression.
For HAP maintainers, reading the text below is not necessary, as it only addresses syntax errors in the user's XPath expression.

Contrary to your claim that no exception is being thrown, the `SelectNodes` method of HtmlAgilityPack 1.11.53 will actually throw an exception with a syntactically incorrect XPath expression like the one in your report. Please pay closer attention when troubleshooting and debugging your projects.

The first issue of your XPath expression is that that it doesn't select any element node on which the `[...]` predicate could be applied. `//[...]` is incorrect syntax. `//` is not a valid selector (or node test). It is an alias for `/descendant-or-self::node()/`; note the trailing slash. Therefore `//[...]` needs to be turned into `//*[...]` to select any nodes.

The second issue is quite similar. `@[...]` is also invalid syntax. `@` is an alias for `attribute::`, and `@[...]` is therefore actually `attribute::[...]` which doesn't make any sense. As with the first issue, your XPath needs to actually select some attribute nodes the _starts-with_ predicate can be applied to. Therefore `@[...]` should here be changed to `@*[...]`, which would apply the predicate to every attribute of each of the nodes selected with `//*`.

But even if you fix those two issues, due to the HAP bug i highlighted above, the XPath expression would still not succeed as it uses the `name()` function, and due to the bug in HAP `name()` does not return attribute names. But the `local-name()` function will do, so using this instead of `name()` can serve as a workaround until the `name()` bug in HAP is being fixed.

In summary, applying the two corrections of invalid XPath syntax and the `local-name()` workaround, you should use the XPath expression:
```
//*[@*[starts-with(local-name(), 'on')]]
```

**Comment by Kenshinofkin** [comment]
@elgonzo this was mistyping. I was trying to quickly create an example using the dotnetfiddle.net and was running into issues. The xpath I used was `//*[@*[starts-with(name(), 'on')]]`. I created a stackoverflow question before hand because I am not familiar with xpath. And someone answered that it was a bug with HAP.

**Comment by JonathanMagnan** [comment]
Stack Overflow issue for reference: https://stackoverflow.com/questions/77204747/html-agility-pack-not-matching-xpath-where-attribute-name-starts-with

**Comment by Kenshinofkin** [comment]
@elgonzo It wasn't a waste. At least not on my end. I did learn a couple of things about xpath and System.Xml.XPath being limited to version XPath 1.0. Thank you,
